### PR TITLE
Changes to allow overriding singletons

### PR DESCRIPTION
**Pull request by dardanmulaj**
Updated the README.md to show an example of how to implement a custom handler like "ID_slug_here" instead of the default "slug-here-ID".

Diving into the example I noticed the `DeferrableProvider` interface causing issues if you're not re-registering the `Rerouter` class in the new service provider. So, despite changing the `BaseSlugSanitizer` separator to `_`, it was still sanitizing using `-`. For now I removed the DeferrableProvider interface from the service provider. You could keep it, but it would require you to register the Rerouter again in the new service provider so it follows the correct order. At least from my testing.

<details>
  <summary>Describing the faulty process</summary>

Basically if you're not re-registering the Rerouter in the new service provider, it would provide the incorrect `$actualValue` parameter value in the `resolveRouteBinding()` function called from the `HasSelfHealingUrls` trait via the second parameter `$model->getRouteKey()`in the reroute function of the Rerouter. So basically it would ignore the new BaseSlugSanitizer separator `_` and use the default hyphen `-` for the route key.

</details>

Just copy+pasted and changed the original tests to match the new example. In the future of course it'd be better to combine the tests somehow to avoid duplicate code - skipped this for now.

### CODE REVIEW

**Comment by lukeraymonddowning** [comment]
Hey! Thanks for the work here. I think I've fixed this issue by using `singletonIf` instead of `singleton` in the service provider. If I've missed the point of your PR, feel free to reopen.
